lib/Filesystem/Domain/FileList.php

```diff
@@ -86,30 +86,51 @@ public function byExtensions(array $extensions): self
     }
 
     /**
-     * @param string[] $globPatterns
+     * @param string[] $includePatterns
+     * @param string[] $excludePatterns
      */
-    public function excludePatterns(array $globPatterns): self
+    public function includeAndExclude(array $includePatterns = [], array $excludePatterns = []): self
     {
-        return $this->filter(function (SplFileInfo $info) use ($globPatterns) {
-            foreach ($globPatterns as $pattern) {
-                if (Glob::match($info->getPathname(), $pattern)) {
-                    return false;
-                }
+        $inclusionMap = [];
+        if ($includePatterns === []) {
+            $inclusionMap['/**/*'] = true;
+        }
+
+        foreach ($includePatterns as $includePattern) {
+            $inclusionMap[$includePattern] = true;
+        }
+        foreach ($excludePatterns as $excludePattern) {
+            $inclusionMap[$excludePattern] = false;
+        }
+
+        // Sort map by keys so that more specific paths are getting matched first
+        uksort($inclusionMap, function (string $a, string $b) {
+            $partsA = explode(DIRECTORY_SEPARATOR, $a);
+            $partsB = explode(DIRECTORY_SEPARATOR, $b);
+            $countDiff = count($partsA) <=> count($partsB);
+            if ($countDiff !== 0) {
```

Comment:
this is the common case or? why not `strlen` compare before exploding instead? should be more performant.